File: backend/app/core/ip_address.py
"""Canonical identity for supported, unscoped IP address literals."""

from ipaddress import ip_address


def canonical_ip_address(value: str) -> str:
    """Parse a complete literal without conflating address families or scopes."""
    text = value.strip()
    if "%" in text:
        raise ValueError("Please enter a valid IPv4 or IPv6 address without a scope")
    try:
        return str(ip_address(text))
    except ValueError as error:
        raise ValueError("Please enter a valid IPv4 or IPv6 address") from error

File: frontend/src/utils/ipAddress.js
/** Validate complete literals; the backend owns canonical persistence. */
export const ipAddressRule = (value) => {
  if (!value) return 'IP address is required'
  const invalid = 'Please enter a valid IPv4 or IPv6 address'
  const text = value.trim()
  if (text.includes(':')) {
    if (!/^[0-9a-f:.]+$/i.test(text)) return invalid
    try {
      // The browser URL parser validates an entire bracketed IPv6 literal.
      return new URL(`http://[${text}]/`).hostname.startsWith('[') || invalid
    } catch {
      return invalid
    }
  }
  const parts = text.split('.')
  return (
    (parts.length === 4 &&
      parts.every((part) => /^(0|[1-9][0-9]{0,2})$/.test(part) && Number(part) <= 255)) ||
    invalid
  )
}
